hostapd: fix logging of configuration content

ckorber opened the pull request:
As discussed in #17517, there are contents of hostapd's configuration file logged in syslog. This includes critical information like passphrase`. To circumvent this condition, this commit logs only "inline" if config_fname is inline data.

Otherwise the upstream logic of hostapd applies.

Fixes: #14049

guidosarducci commented:
@Chris1189 Hi Christian, I tested this locally and can confirm it works as expected.

I also looked at the various log-level outputs and docs, and recommend changing `MSG_MSGDUMP` to `MSG_DEBUG` as I think you originally had, especially since `DEBUG` is a standard syslog level and will be less confusing for users.

Please add a "Fixes:" tag to your commit message mentioning https://github.com/openwrt/openwrt/issues/14049, so this might get some attention. I've also requested that issue be flagged for the **24.10** release, so hopefully this PR does too.

FYI, see also this old issue (still present) regarding setting hostapd log levels: https://github.com/openwrt/openwrt/issues/8061

guidosarducci commented on package/network/services/hostapd/patches/701-reload_config_inline.patch:
```diff
@@ -39,3 +39,20 @@ as adding/removing interfaces.
+ 
+-	wpa_printf(MSG_INFO, "Configuration file: %s (phy %s)%s",
++	if (!strncmp(config_fname, "data:", 5)) {
++		wpa_printf(MSG_MSGDUMP, "Configuration file: %s (phy %s)%s",
```
Suggest using `MSG_DEBUG` instead of `MSG_MSGDUMP` but otherwise this works.

KanjiMonster replied:
I don't like that it removes any output in this case when not enabling DEBUG, so I suggest doing an additional MSG_INFO message with "<inline>" as the filename. That way the log line will always appear, regardless of using inline or not, and then only with MSG_DEBUG also print the contents.

Though I'm unsure if we should print the contents here at all. At least it feels like the wrong place for that, I would rather expect this in "{hostapd,wpa}_config_read()".

Also there are multiple further places where the filename is printed (e.g. `hostapd_interface_init()`) where it will print it in an error message, these should also be changed to not print the contents.

guidosarducci replied:
> I don't like that it removes any output in this case when not enabling DEBUG, so I suggest doing an additional MSG_INFO message with "" as the filename. That way the log line will always appear, regardless of using inline or not, and then only with MSG_DEBUG also print the contents.

Agreed. Something like "Configuration: \<inline\>" would be consistent.
> Though I'm unsure if we should print the contents here at all. At least it feels like the wrong place for that, I would rather expect this in "{hostapd,wpa}_config_read()".

Again agreed. Dumping the contents feels like a kludge. IIRC, the contents are appended in some ucode? That would be a better place. However, I had no response from ucode or hostapd experts here or on IRC since first raising this issue more than a year ago, so I'm fine with resolving it like the PR. Perhaps there's a simpler way that @nbd or @jow- can suggest?
 
> Also there are multiple further places where the filename is printed (e.g. `hostapd_interface_init()`) where it will print it in an error message, these should also be changed to not print the contents.

Interesting. Are you sure the filename var has been appended at these points? Dealing with the filename var at the point it's built up would simplify all these cases.

Neustradamus commented:
Interesting PR :)

Commit:
hostapd: fix logging of configuration content

As discussed in openwrt#17517, there are contents of hostapd's configuration file logged in syslog.
This includes critical information like `passphrase`. To circumvent this condition,
this commit logs only "inline" if config_fname is inline data.

Otherwise the upstream logic of hostapd applies.

Fixes: openwrt#14049

Signed-off-by: Christian Korber <ck@dev.tdt.de>
Link: https://github.com/openwrt/openwrt/pull/17718
Signed-off-by: Robert Marko <robimarko@gmail.com>

ckorber commented on package/network/services/hostapd/patches/701-reload_config_inline.patch:
```diff
@@ -8,34 +8,75 @@ as adding/removing interfaces.
++		wpa_printf(MSG_INFO, "Configuration file: %s",
++			name);
 +	}
  	wpa_printf(MSG_DEBUG, "Reading configuration file '%s'", name);
```
Ah I know what you mean! I dropped the `wpa_printf()` and changed the output as you said.

ckorber commented on package/network/services/hostapd/patches/701-reload_config_inline.patch:
```diff
@@ -8,34 +8,75 @@ as adding/removing interfaces.
  	size_t i;
  
 -	f = fopen(fname, "r");
 +	if (!strncmp(fname, "data:", 5)) {
 +		f = fmemopen((void *)(fname + 5), strlen(fname + 5), "r");
 +		fname = "<inline>";
++		wpa_printf(MSG_INFO, "Configuration file: %s",
++			fname);
 +	} else {
 +		f = fopen(fname, "r");
++		wpa_printf(MSG_INFO, "Configuration file: %s",
++			fname);
 +	}
```
I made the changes. Thank you for review!

KanjiMonster replied:
What I meant is that you move the `wpa_printf() out of if-else, since its the same in both, and just have once.

And change the message to the same already in use by `wpa_config_read()`, which is "`"Reading configuration file '%s'"`".

And finally (but this was more a suggestion), use `MSG_DEBUG here` as well since wpa_config_read() also uses `MSG_DEBUG`, as well as `hostapd_interface_init()` and `hostapd_interface_init_bss()` do).

ckorber replied:
I think now I got your point. I moved it out of the block.
Personally, I think `MSG_INFO` is also fine as information. Whether it is `MSG_INFO` or `MSG_DEBUG`, both would be okay with me.

ckorber commented:
I tested it again after making the changes @KanjiMonster proposed and it works.
The output is now reduced to one line:
```
Fri May 23 06:09:04 2025 daemon.notice hostapd: Configuration file: <inline>
```

KanjiMonster commented on package/network/services/hostapd/patches/701-reload_config_inline.patch:
```diff
@@ -28,14 +32,48 @@ as adding/removing interfaces.
+-	f = fopen(name, "r");
 +	if (!strncmp(name, "data:", 5)) {
 +		f = fmemopen((void *)(name + 5), strlen(name + 5), "r");
 +		name = "<inline>";
 +	} else {
 +		f = fopen(name, "r");
 +	}
- 	wpa_printf(MSG_DEBUG, "Reading configuration file '%s'", name);
--	f = fopen(name, "r");
++	wpa_printf(MSG_DEBUG, "%s --> Reading configuration file ", name);
```
I think there was a bit of a miscommuniation here (or a copy-paste mistake), I meant to keep the original print in place here.

ckorber replied:
I reverted it to the last sentence.

KanjiMonster commented on package/network/services/hostapd/patches/701-reload_config_inline.patch:
```diff
@@ -28,14 +32,48 @@ as adding/removing interfaces.
+ 		}
+ 	}
+ 
+-	wpa_printf(MSG_INFO, "Configuration file: %s (phy %s)%s",
+-		   config_fname, phy, iface ? "" : " --> new PHY");
++	if (!strncmp(config_fname, "data:", 5)) {
++		wpa_printf(MSG_DEBUG, "Configuration file: %s (phy %s)%s",
++			   "<inline>", phy, iface ? "" : " --> new PHY");
++	} else {
++		wpa_printf(MSG_DEBUG, "Configuration file: %s (phy %s)%s",
++			   config_fname, phy, iface ? "" : " --> new PHY");
++	}
```
Just saw it, if the original was `MSG_INFO`, then keep it at `MSG_INFO`.
```suggestion
-	wpa_printf(MSG_INFO, "Configuration file: %s (phy %s)%s",
-		   config_fname, phy, iface ? "" : " --> new PHY");
+	if (!strncmp(config_fname, "data:", 5)) {
+		wpa_printf(MSG_INFO, "Configuration file: %s (phy %s)%s",
+			   "<inline>", phy, iface ? "" : " --> new PHY");
+	} else {
+		wpa_printf(MSG_INFO, "Configuration file: %s (phy %s)%s",
+			   config_fname, phy, iface ? "" : " --> new PHY");
+	}
```

ckorber replied:
Done

KanjiMonster commented on package/network/services/hostapd/patches/701-reload_config_inline.patch:
```diff
@@ -36,6 +39,40 @@ as adding/removing interfaces.
 +	}
  	wpa_printf(MSG_DEBUG, "Reading configuration file '%s'", name);
 -	f = fopen(name, "r");
++
```
Now you are adding a stray empty line here. Drop this, and then this is perfect (IMHO)

ckorber replied:
Okay. Removed the line!

KanjiMonster replied:
You removed a different empty line! :laughing:

This one is still there/being added (in `wpa_config_read()`).

ckorber replied:
Not my day :laughing:

ckorber replied:
Now it should be free of blank lines.

KanjiMonster reviewed (approved):
Now it looks good!

Neustradamus commented:
@KanjiMonster: Thanks for your review, and @ckorber for changes!

Important: After merging in main branch, it must be backported in previous OpenWrt versions.

ckorber commented:
The PR ist still pending without approvals. Maybe a review from you @nbd168 ?

Would appreciate your help :grin:

ckorber commented:
Is there anything open from your side @nbd168 ? Or any advice how to get the needed approvals?
Thank you in advance!

robimarko commented:
Thanks! Rebased on top of main and merged!
